sys/kerncore/src/lib.rs

```diff
@@ -159,35 +178,53 @@ where
 
     // Per the function's preconditions, the region table is sorted in ascending
     // order of base address, and the regions within it do not overlap. This
-    // lets us use a one-pass algorithm.
+    // lets us use a binary search followed by a short scan
     let mut scan_addr = slice.base_addr();
     let end_addr = slice.end_addr();
 
-    for region in table {
-        if region.contains(scan_addr) {
-            // Make sure it's permissible!
-            if !region_ok(region) {
-                // bail to the fail handling code at the end.
-                break;
-            }
-
-            if end_addr <= region.end_addr() {
-                // We've exhausted the slice in this region, we don't have
-                // to continue processing.
-                return true;
-            }
+    let Ok(index) =
+        table.binary_search_by(|reg| region_compare(reg, scan_addr))
+    else {
+        // No region contained the start address.
+        return false;
+    };
+
+    // Perform fast checks on the initial region. In practical testing this
+    // provides a ~1% performance improvement over only using the loop below.
+    let first_region = &table[index];
+    if !region_ok(first_region) {
+        return false;
+    }
+    // Advance to the end of the first region
+    scan_addr = first_region.end_addr();
+    if scan_addr >= end_addr {
+        // That was easy
+        return true;
+    }
 
-            // Continue scanning at the end of this region.
-            scan_addr = region.end_addr();
-        } else if region.base_addr() > scan_addr {
-            // We've passed our target address without finding regions that
-            // work!
+    // Scan adjacent regions.
+    for region in &table[index + 1..] {
+        if !region.contains(scan_addr) {
+            // We've hit a hole without finishing our scan.
+            break;
+        }
+        // Make sure the region is permissible!
+        if !region_ok(region) {
+            // bail to the fail handling code at the end.
             break;
```

Comment:
hum, the "fail handling code at the end" is just `false`...so couldn't these just be `return false;`? I suppose it makes  sense to have them not be if we anticipate probably adding additional code that runs when no valid region is found, but...I dunno. Not a big deal, as the current thing is fine.

Reply:
:woman_shrugging: well, it's not terribly important to me --- i just went hunting for what the fault handling code at the end might be, and then, when I saw that it was just `false`, kind of went, "wait, what am I missing here?" for a bit.